src/components/ebay-table/component.ts

```diff
@@ -11,13 +19,119 @@ export interface TableCell
 
-export default class extends Marko.Component<Input> {}
+enum RowState {
+    SELECTED = "selected",
+    DISABLED = "disabled",
+}
+
+interface State extends Record<RowState, TableRowStateMapping> {}
+
+export default class extends Marko.Component<Input, State> {
+    onCreate(input: Input) {
+        this.state = {
+            ...this.getRowStateFromInput(input),
+        };
+    }
+
+    repeatableAttrTagToArr<T>(
+        attr?: Marko.RepeatableAttrTag<T> | Marko.AttrTag<T>[],
+    ): T[] {
+        return Array.isArray(attr) ? attr : attr ? [attr] : [];
+    }
+
+    getRowStateFromInput(input: Input) {
+        const selected = { ...input.selected };
+        const disabled: TableRowStateMapping = {};
+        this.repeatableAttrTagToArr(input.row).forEach((row, i) => {
+            const name = row.name || i;
+            selected[name] ??= false;
+            disabled[name] = !!row.disabled;
+        });
+        return { selected, disabled };
+    }
+
+    getDisabledFromInput(input: Input): TableRowStateMapping {
+        const disabled: TableRowStateMapping = {};
+        this.repeatableAttrTagToArr(input.row).forEach((row) => {
+            if (row.disabled) {
+                disabled[row.name || ""] = true;
+            }
+        });
+        return disabled;
+    }
+
+    getSelectionStatus(
+        selected: TableRowStateMapping = this.state.selected,
+    ): AttrTriState {
+        let selectedCount = 0;
+        let rowCount = 0;
+        for (const [name, isSelected] of Object.entries(selected)) {
+            if (this.state.disabled[name]) {
+                continue;
+            }
+            if (isSelected) {
+                selectedCount++;
+            }
+            rowCount++;
+        }
+        if (selectedCount === 0) {
+            return "false";
+        }
+        if (selectedCount === rowCount) {
+            return "true";
+        }
+        return "mixed";
+    }
+
+    headerSelect() {
+        const selectedRenderedRowState = Object.keys(
+            this.state.disabled,
+        ).reduce((acc, name) => {
+            acc[name] = this.state.selected[name];
+            return acc;
+        }, {} as TableRowStateMapping);
+        const renderedSelectionStatus = this.getSelectionStatus(
+            selectedRenderedRowState,
+        );
+        this.state.selected = {
+            ...this.state.selected,
+            ...Object.keys(selectedRenderedRowState).reduce((acc, name) => {
+                if (!this.state.disabled[name]) {
+                    acc[name] = renderedSelectionStatus !== "true";
+                }
+                return acc;
+            }, {} as TableRowStateMapping),
+        };
+        this.emit("select", {
+            selected: this.state.selected,
+        });
+    }
+
+    rowSelect(name: TableRowName, { checked }: CheckboxEvent) {
+        this.state.selected[name] = checked;
+        this.setStateDirty(RowState.SELECTED);
```

Comment:
This only works because `RowState.SELECTED` is `"selected"`. You should
1. Change the previous line to `this.state[RowState.SELECTED][name] = checked`
2. Change this line to `this.setStateDirty("selected")`
3. Duplicate the array without mutating so you don't have to `setStateDirty`

Reply:
I was doing option 1, but somehow typescript compiler allows `this.state.disabled`, thought it is an acceptable pattern. I'll change it back if not preferred.